File: src/main/java/ru/xpendence/topjavagraduation/entity/Review.java
package ru.xpendence.topjavagraduation.entity;

import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;

import java.time.LocalDate;

@Entity
@Table(
        name = "reviews",
        uniqueConstraints = @UniqueConstraint(
                name = "ux_user_restaurant",
                columnNames = {"user_id", "restaurant_id"}
        )
)
@Getter
@Setter
public class Review {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "rating", nullable = false)
    private Integer rating;

    @Column(name = "text", length = 1000)
    private String text;

    @Column(name = "date", nullable = false)
    private LocalDate date;

    @ManyToOne(fetch = FetchType.LAZY, cascade = CascadeType.DETACH)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @ManyToOne(fetch = FetchType.LAZY, cascade = CascadeType.DETACH)
    @JoinColumn(name = "restaurant_id", nullable = false)
    private Restaurant restaurant;

    public static void enrichForUpdate(Review forUpdate, Review stored) {
        stored.rating = forUpdate.rating;
        stored.text = forUpdate.text;
    }
}

File: src/main/kotlin/ru/xpendence/topjavagraduation/entity/Review.kt
package ru.xpendence.topjavagraduation.entity

import java.time.LocalDate

/**
 * Отзыв.
 */
data class Review(
    val id: Long? = null,
    val rating: Int,
    val text: String? = null,
    val date: LocalDate,
    val user: User,
    val restaurant: Restaurant,
) {
    companion object {
        @JvmStatic
        fun enrichForUpdate(forUpdate: Review, stored: Review): Review =
            stored.copy(
                rating = forUpdate.rating,
                text = forUpdate.text,
            )
    }
}
